packages/editor-preview/src/browser/editor-preview-widget.ts

```diff
@@ -0,0 +1,226 @@
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import {
+    ApplicationShell, BaseWidget, Navigatable, OpenerOptions, PanelLayout, Saveable,
+    WidgetConstructionOptions, StatefulWidget, Widget, WidgetManager, DockPanel, Title
+} from '@theia/core/lib/browser';
+import { Emitter, DisposableCollection } from '@theia/core/lib/common';
+import URI from '@theia/core/lib/common/uri';
+import { EditorWidget } from '@theia/editor/lib/browser';
+import { Message } from '@phosphor/messaging';
+import { find } from '@phosphor/algorithm';
+
+export interface PreviewEditorOpenerOptions extends OpenerOptions  {
+    preview?: boolean
+}
+
+export interface PreviewViewState {
+    pinned: boolean,
+    editorState: object,
+    previewDescription: WidgetConstructionOptions
+}
+
+export interface PreviewEditorPinnedEvent {
+    preview: EditorPreviewWidget,
+    editorWidget: EditorWidget
+}
+
+/** The class name added to Editor Preview Widget titles. */
+const PREVIEW_TITLE_CLASS = ' theia-editor-preview-title-unpinned';
+
+export class EditorPreviewWidget extends BaseWidget implements ApplicationShell.TrackableWidgetProvider, Navigatable, StatefulWidget {
+
+    protected pinned_: boolean;
+    protected pinListeners = new DisposableCollection();
+    protected onDidChangeTrackableWidgetsEmitter = new Emitter<Widget[]>();
+
+    readonly onDidChangeTrackableWidgets = this.onDidChangeTrackableWidgetsEmitter.event;
+
+    protected onPinnedEmitter = new Emitter<PreviewEditorPinnedEvent>();
+
+    readonly onPinned = this.onPinnedEmitter.event;
+
+    constructor(protected widgetManager: WidgetManager, protected editorWidget_?: EditorWidget) {
+        super();
+        this.title.closable = true;
+        this.title.className += PREVIEW_TITLE_CLASS;
+        this.layout = new PanelLayout();
+        this.toDispose.push(this.onDidChangeTrackableWidgetsEmitter);
+        this.toDispose.push(this.onPinnedEmitter);
+        this.toDispose.push(this.pinListeners);
+    }
+
+    get editorWidget() {
+        return this.editorWidget_;
+    }
+
+    get pinned() {
+        return this.pinned_;
+    }
+
+    get saveable(): Saveable|undefined {
+        if (this.editorWidget_) {
+            return this.editorWidget_!.saveable;
+        }
+    }
+
+    getResourceUri(): URI | undefined {
+        return this.editorWidget_!.getResourceUri();
+    }
+    createMoveToUri(resourceUri: URI): URI | undefined {
+        return this.editorWidget_!.createMoveToUri(resourceUri);
+    }
+
+    pinEditorWidget() {
```

Comment:
please put return types explicitly, even when they are  `void` to avoid accidental breaking changes